ejb3/src/main/java/org/jboss/as/ejb3/timerservice/TimerServiceImpl.java

```diff
@@ -594,10 +594,48 @@ public void persistTimer(final TimerImpl timer, boolean newTimer) {
                     EJB3_TIMER_LOGGER.timerPersistenceNotEnable();
                     return;
                 }
+
+                final ContextTransactionManager transactionManager = ContextTransactionManager.getInstance();
+                Transaction clientTX = transactionManager.getTransaction();
                 if (newTimer) {
-                    timerPersistence.getValue().addTimer(timer);
+                    if( clientTX == null ){
+                        transactionManager.begin();
+                    }
+                    try {
+                        timerPersistence.getValue().addTimer(timer);
+                        if(clientTX == null) transactionManager.commit();
+                    } catch (Exception e){
+                        if(clientTX == null) {
+                            try {
+                                transactionManager.rollback();
+                            } catch (Exception ee){
+                                EjbLogger.EJB3_TIMER_LOGGER.timerUpdateFailedAndRollbackNotPossible(ee);
+                            }
+                        }
+                        throw e;
+                    }
                 } else {
-                    timerPersistence.getValue().persistTimer(timer);
+                    if( clientTX != null ) transactionManager.suspend();
+                    for (int count = 3; count > 0; count--) {
```

Comment:
The retry could should not be hard coded, at the very least it should be a system property.

Reply:
Again, you are absolutely right! This is ugly.

But this is the most significant and vulnerable point of the whole process. If we will not be able to finish this persistence - the Timer will never be invoked again. And this is the problem, which I have to eliminate (especially for the Oracle DB with its SQLExceptions on any concurrent modifications).

Maybe, there is some other way how to make repeats. I just don't familiar enough with the project to perform it right or make it configurable. Sorry. 

It just an idea, which eliminates the huge amount of "headache" in my project.

Reply:
Hi @stuartwdouglas !
I've made changes in this code. I hope that's will be alright. Please, see my last commits in this request